### PR TITLE
Fetch refactor

### PR DESCRIPTION
**Pull request by max-hoffman**
- consolidate `getRefSpecs` functions from cli/env
- fetch argument switching matches git more closely -- specifying the remote is necessary, no default to origin

### CODE REVIEW

File: go/libraries/doltcore/env/remotes.go [hunk 1]
```diff
@@ -241,20 +241,18 @@ func NewFetchOpts(args []string, rsr RepoStateReader) (Remote, []ref.RemoteRefSp
...
-			remoteOK = true
-			args = args[1:]
-		}
+	var remName string
+	if len(args) == 0 {
+		remName = "origin"
+	} else {
+		remName = args[0]
+		args = args[1:]
 	}
```
Comment (max-hoffman) [inline go/libraries/doltcore/env/remotes.go:250]: if there is a first argument, must be valid remote (like git)

File: go/cmd/dolt/commands/fetch.go [hunk 2]
```diff
@@ -77,164 +76,15 @@ func (cmd FetchCmd) Exec(ctx context.Context, commandStr string, args []string,
...
-			args = args[1:]
-		}
-	}
-
-	if !remoteOK {
-		return env.NoRemote, nil, errhand.BuildDError("error: unknown remote").SetPrintUsage().Build()
-	}
-
-	var rs []ref.RemoteRefSpec
 	var verr errhand.VerboseError
```
Comment (zachmu) [inline go/cmd/dolt/commands/fetch.go:79]: In general, this is an anti-pattern: very easy to just not handle an error when it's declared outside where it's used.

Handle each branch's error separately, returning immediately.

File: go/libraries/doltcore/env/actions/remotes.go [hunk 3]
```diff
@@ -358,3 +358,68 @@ func FetchRemoteBranch(ctx context.Context, tempTablesDir string, rem env.Remote
 
 	return srcDBCommit, nil
 }
+
+func FetchRefSpecs(ctx context.Context, dbData env.DbData, refSpecs []ref.RemoteRefSpec, remote env.Remote, mode ref.UpdateMode, progStarter ProgStarter, progStopper ProgStopper) error {
```
Comment (zachmu) [inline go/libraries/doltcore/env/actions/remotes.go:362]: Method docs please

File: go/libraries/doltcore/env/actions/remotes.go [hunk 4]
```diff
@@ -358,3 +358,68 @@ func FetchRemoteBranch(ctx context.Context, tempTablesDir string, rem env.Remote
...
+			if remoteTrackRef != nil {
+				rsSeen = true
+				srcDBCommit, err := FetchRemoteBranch(ctx, dbData.Rsw.TempTableFilesDir(), remote, srcDB, dbData.Ddb, branchRef, remoteTrackRef, progStarter, progStopper)
+				if err != nil {
+					return err
+				}
+
+				switch mode {
+				case ref.ForceUpdate:
+					err := dbData.Ddb.SetHeadToCommit(ctx, remoteTrackRef, srcDBCommit)
```
Comment (zachmu) [inline go/libraries/doltcore/env/actions/remotes.go:388]: This is unfortunate, should add a TODO here (can't be used safely in SQL context)

File: go/libraries/doltcore/env/actions/remotes.go [hunk 5]
```diff
@@ -358,3 +358,68 @@ func FetchRemoteBranch(ctx context.Context, tempTablesDir string, rem env.Remote
...
+						return fmt.Errorf("%w: %s", ErrCantFF, err.Error())
+					}
+					if !ok {
+						return ErrCantFF
+					}
+
+					switch err {
+					case doltdb.ErrUpToDate:
+					case doltdb.ErrIsAhead, nil:
+						err = dbData.Ddb.FastForward(ctx, remoteTrackRef, srcDBCommit)
```
Comment (zachmu) [inline go/libraries/doltcore/env/actions/remotes.go:404]: Same issue here

Comment (zachmu) [reply go/libraries/doltcore/env/actions/remotes.go:404]: This is fine, just needs to be documented and we need to come up with a reasonable set of behaviors for open connections

File: integration-tests/bats/sql-fetch.bats [hunk 6]
```diff
@@ -217,7 +217,7 @@ teardown() {
 @test "sql-fetch: dolt_fetch unknown remote fails" {
     cd repo2
     dolt remote remove origin
-    run dolt sql -q "select dolt_fetch('unknown', 'master')"
+    run dolt sql -q "select dolt_fetch('unknown')"
```
Comment (zachmu) [inline integration-tests/bats/sql-fetch.bats:220]: Is this not a valid test?

**Review by zachmu — APPROVED** [review]
Looks fine, I like the refactor

File: integration-tests/bats/sql-fetch.bats (hunk 6 above)
Comment (max-hoffman) [reply integration-tests/bats/sql-fetch.bats:220]: both are valid, but remotes are resolved before refspecs, so i thought the addition of a branch name distracted from what was actually being tested

File: go/libraries/doltcore/env/actions/remotes.go (hunk 4 above)
Comment (max-hoffman) [reply go/libraries/doltcore/env/actions/remotes.go:388]: lots of server locking todos

**Pull request merged**